### PR TITLE
Add wildcard matches for virtual hosts

### LINKED ISSUE #519: router: support wildcard matches
It should be possible to match *.foo.com and *-suffix.foo.com.

### PR DESCRIPTION
**Pull request by tschroed**
* is already supported. This adds *.foo.com and *-bar.foo.com. 

fixes #519

### CODE REVIEW

**Comment by mattklein123** [comment]
Quick note that we will need docs for this change (at minimum here: https://lyft.github.io/envoy/docs/configuration/http_conn_man/route_config/vhost.html)

File: source/common/common/trie.h [hunk 1]
```diff
@@ -0,0 +1,111 @@
+#pragma once
+
+#include <memory>
```
Comment (mattklein123) [inline source/common/common/trie.h]: nit: del standard includes like these and add to precompiled.h if not already there. Same applies elsewhere.

File: source/common/common/trie.h [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+ * matches.
+ *
+ * Note that Value must have a conversion constructor that will take nullptr.
+ */
+// TODO(tschroed): It's possible for Key to be a different type from the
+// internal path component. E.g. a uint32_t IPv4 prefix can be tokenized to
+// uint8_t components. Likewise a string representation of same.
+template <class Key, class Value, class Tokenizer> class TrieNode {
+public:
+  TrieNode(Value value) : value_(value) {}
```
Comment (mattklein123) [inline source/common/common/trie.h]: We can add a move constructor later if needed, but for now I would do const Value&

File: source/common/common/trie.h [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+  TrieNode() : TrieNode(nullptr) {}
+  ~TrieNode() {}
+
+  /**
+   * Insert a value into the trie.
+   * @param key the name associated with the value.
+   * @param value the value to store on the node associated with name
+   */
+  void emplace(const Key& key, Value value);
```
Comment (mattklein123) [inline source/common/common/trie.h]: again per above, can add move later, but I would do const Value& for now.

File: source/common/common/trie.h [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+  /**
+   * Lookup a value by key.
+   * @param key the path name to look up
+   * @return a std::pair consisting of the value if found and a boolean
+   *         indicating if the match was exact.
+   */
+  std::pair<Value, bool> find(const Key& key) const;
```
Comment (mattklein123) [inline source/common/common/trie.h]: I agree with @htuch here. I don't see the need to copy here. I would return const Value*, or std::reference_wrapper<const Value> as part of the pair.

File: source/common/common/trie.h [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+private:
+  std::pair<Value, bool> find(std::vector<Key>& path_components) const;
+  void emplace(std::vector<Key>& path_components, Value value);
+  Value value() const { return value_; }
+  void set_value(Value value) { value_ = value; }
```
Comment (mattklein123) [inline source/common/common/trie.h]: const Value&

File: source/common/common/trie.h [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+  // incomplete types.
+  std::unordered_map<Key, std::unique_ptr<TrieNode<Key, Value, Tokenizer>>> children_;
+  Value value_;
+  Tokenizer tokenizer_;
+};
+
+template <class Key, class Value, class Tokenizer>
+void TrieNode<Key, Value, Tokenizer>::emplace(const Key& key, Value value) {
+  std::vector<Key> path_components = tokenizer_.tokenize(key);
+  // Keys come in big endian but we reverse them for vector iteration
```
Comment (mattklein123) [inline source/common/common/trie.h]: This comment is a little strange regarding endian, if I understand you reverse here so that you can pop_back later, I would just say that.

File: source/common/common/trie.h [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+void TrieNode<Key, Value, Tokenizer>::emplace(std::vector<Key>& path_components, Value value) {
+  if (path_components.empty()) {
+    return;
+  }
+  Key name(path_components.back());
+  path_components.pop_back();
+  if (children_.find(name) == children_.end()) {
+    std::unique_ptr<TrieNode<Key, Value, Tokenizer>> node(
+        new TrieNode<Key, Value, Tokenizer>(nullptr));
+    children_[name] = std::move(node);
```
Comment (mattklein123) [inline source/common/common/trie.h]: nit: You can do this in one line via children_[name].reset(...)

Better, you can avoid 2 hash lookups via this trick:
std::unique_ptr<TrieNode<Key, Value, Tokenizer>>& node = children_[name];
if (!node) {
  node.reset(...)
}

File: source/common/common/trie.h [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+    node->set_value(value);
+  } else {
+    node->emplace(path_components, value);
+  }
+}
+
+template <class Key, class Value, class Tokenizer>
+std::pair<Value, bool> TrieNode<Key, Value, Tokenizer>::find(const Key& key) const {
+  std::vector<Key> path_components = tokenizer_.tokenize(key);
+  // Keys come in big endian but we reverse them for vector iteration
```
Comment (mattklein123) [inline source/common/common/trie.h]: same comment here about comment

Comment (mattklein123) [inline source/common/common/trie.h]: might just have helper to tokenize and reverse with comment

File: source/common/common/trie.h [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+    // our path traversal.
+    return std::make_pair(value_, value_ ? true : false);
+  }
+  bool exact_match = false;
+  Value value = value_;
+  Key component(path_components.back());
+  path_components.pop_back();
+  if (children_.find(component) != children_.end()) {
+    const TrieNode<Key, Value, Tokenizer>* node = children_.find(component)->second.get();
+    std::pair<Value, bool> retval = node->find(path_components);
```
Comment (mattklein123) [inline source/common/common/trie.h]: Optimally, this function would be written without recursion, in case we have deep nesting. I don't think it matters that much but I would drop a comment in.

File: source/common/common/utility.cc [hunk 10]
```diff
@@ -102,6 +102,30 @@ std::string StringUtil::subspan(const std::string& source, size_t start, size_t
   return source.substr(start, end - start);
 }
 
+std::vector<std::string> StringUtil::tokenize(const std::string& source,
```
Comment (mattklein123) [inline source/common/common/utility.cc]: This is a more generalized version of StringUtil::split(). I'm not sure if we really need both? Should we get rid of split() and just call tokenize split and make it take multiple splitters? This one will be a bit slower, but not sure if it's worth having both versions.

Either way, you need test cases for this function in utility_test.cc.

File: test/common/common/trie_test.cc [hunk 11]
```diff
@@ -0,0 +1,161 @@
+#include <vector>
+
+#include <algorithm>
+#include "common/common/trie.h"
+#include "common/common/utility.h"
+
+namespace {
+
+struct ReverseDomainTokenizer {
+  std::vector<std::string> tokenize(std::string str) const {
```
Comment (mattklein123) [inline test/common/common/trie_test.cc]: const std::string& str

File: source/common/router/config_impl.h [hunk 12]
```diff
@@ -1,15 +1,26 @@
...
 #include "envoy/runtime/runtime.h"
 #include "envoy/upstream/cluster_manager.h"
 
+#include "common/common/trie.h"
+#include "common/common/utility.h"
 #include "common/router/config_utility.h"
 
 namespace Router {
+struct ReverseDomainTokenizer {
+  std::vector<std::string> tokenize(std::string str) const {
```
Comment (mattklein123) [inline source/common/router/config_impl.h]: const std::string& str

File: source/common/router/config_impl.h [hunk 13]
```diff
@@ -1,15 +1,26 @@
...
 #include "envoy/router/router.h"
 #include "envoy/runtime/runtime.h"
 #include "envoy/upstream/cluster_manager.h"
 
+#include "common/common/trie.h"
+#include "common/common/utility.h"
 #include "common/router/config_utility.h"
 
 namespace Router {
+struct ReverseDomainTokenizer {
```
Comment (mattklein123) [inline source/common/router/config_impl.h]: comment on what this is for.

File: source/common/common/trie.h (hunk 1 above)
Comment (tschroed) [inline source/common/common/trie.h]: Ah, thanks for the pointer. I'll update the contributing doc. Most Googlers have been indoctrinated the opposite way.

File: source/common/common/trie.h (hunk 2 above)
Comment (tschroed) [inline source/common/common/trie.h]: Done

File: source/common/common/trie.h (hunk 6 above)
Comment (tschroed) [inline source/common/common/trie.h]: Done.

File: source/common/common/trie.h (hunk 7 above)
Comment (tschroed) [inline source/common/common/trie.h]: Indeed.

File: source/common/common/trie.h (hunk 8 above)
Comment (tschroed) [inline source/common/common/trie.h]: wfm

File: source/common/common/trie.h (hunk 9 above)
Comment (tschroed) [inline source/common/common/trie.h]: Commented.

File: source/common/common/utility.cc (hunk 10 above)
Comment (tschroed) [inline source/common/common/utility.cc]: Close, but not quite

split("hello-world", '-') => {"hello", "world"}
tokenize("hello-world", "-") => {"hello", "-", "world"}

If you're cool with an additional argument indicating whether the splitters should be retained or discarded, I can merge them. I'd prefer to do it in a follow-on change to keep the new feature distinct from cleanup.

File: test/common/common/trie_test.cc (hunk 11 above)
Comment (tschroed) [inline test/common/common/trie_test.cc]: Done.

File: source/common/router/config_impl.h (hunk 13 above)
Comment (tschroed) [inline source/common/router/config_impl.h]: Commented and const'd.

File: source/common/common/trie.h (hunk 2 above)
Comment (tschroed) [inline source/common/common/trie.h]: Done.

File: source/common/common/trie.h (hunk 3 above)
Comment (tschroed) [inline source/common/common/trie.h]: Sure.

File: source/common/common/trie.h (hunk 4 above)
Comment (tschroed) [inline source/common/common/trie.h]: The reason I'm using value types throughout is because in the general case we can't assume that the elements being stored have been heap allocated.

Comment (tschroed) [inline source/common/common/trie.h]: ... and if the user wants to avoid copying large types, then pointers (as we do in config_impl) are always available.

File: source/common/common/utility.cc (hunk 10 above)
Comment (mattklein123) [inline source/common/common/utility.cc]: OK, fair enough, can you add a few more comments to utility.h that show the difference between split and tokenize? It wasn't clear to me.

File: source/common/common/trie.h (hunk 4 above)
Comment (mattklein123) [inline source/common/common/trie.h]: You are returning a copy of value_ which is contained inside TrieNode. (You copy value from the outside in). It doesn't matter if it was originally heap or stack allocated. I'm not sure I see the point of copying here when you can just return a ```const Value*``` that points to value_ inside the TrieNode.

File: source/common/common/trie.h [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+   * @param key the path name to look up
+   * @return a std::pair consisting of the value if found and a boolean
+   *         indicating if the match was exact.
+   */
+  std::pair<Value, bool> find(const Key& key) const;
+
+private:
+  std::pair<Value, bool> find(std::vector<Key>& path_components) const;
+  void emplace(std::vector<Key>& path_components, const Value& value);
+  Value value() const { return value_; }
```
Comment (mattklein123) [inline source/common/common/trie.h]: I don't think this is used?

Comment (tschroed) [inline source/common/common/trie.h]: It seemed nicely symmetric with set_value when I was initially writing this, but clearly not used. It's trivial to add later if the need arises.

File: source/common/common/trie.h (hunk 4 above)
Comment (tschroed) [inline source/common/common/trie.h]: Err. Yeah, obviously working past my brain-on time. How about returning a non-const pointer since it's entirely plausible that someone would want to perform mutations on the associated value?

Comment (mattklein123) [inline source/common/common/trie.h]: Sure, sounds good. We could add a const method later that returns a const pointer if needed.

File: source/common/common/trie.h [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+                                              const Value& value) {
+  if (path_components.empty()) {
+    return;
+  }
+  Key name(path_components.back());
+  path_components.pop_back();
+  if (children_.find(name) == children_.end()) {
+    std::unique_ptr<TrieNode<Key, Value, Tokenizer>> node(
+        new TrieNode<Key, Value, Tokenizer>(nullptr));
+    children_[name].reset(node.release());
```
Comment (mattklein123) [inline source/common/common/trie.h]: nit: sorry I meant 1 line:

```children_[name].reset(new TrieNode<Key, Value, Tokenizer>(nullptr));```

File: source/common/common/trie.h [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+  if (path_components.empty()) {
+    // An exact match requires that we have a non-false value at the very end of
+    // our path traversal.
+    return std::make_pair(&value_, value_ ? true : false);
+  }
+  bool exact_match = false;
+  Value* value = &value_;
+  Key component(path_components.back());
+  path_components.pop_back();
+  if (children_.find(component) != children_.end()) {
```
Comment (mattklein123) [inline source/common/common/trie.h]: avoid double hash lookup here:

```
auto it = children_.find(component);
if (it != children_.end()) {
... = it->second.get();
}
```

File: source/common/router/config_impl.cc [hunk 17]
```diff
@@ -490,11 +492,11 @@ const VirtualHostImpl* RouteMatcher::findVirtualHost(const Http::HeaderMap& head
...
     return iter->second.get();
-  } else if (default_virtual_host_) {
-    return default_virtual_host_.get();
+  } else if (VirtualHostPtr vhost =
+                 *wildcard_virtual_host_suffixes_.find(headers.Host()->value().c_str()).first) {
+    return vhost.get();
   }
-
-  return nullptr;
+  return default_virtual_host_.get();
```
Comment (mattklein123) [inline source/common/router/config_impl.cc]: why change this line?

File: source/common/common/trie.h (hunk 15 above)
Comment (tschroed) [inline source/common/common/trie.h]: I actually reworked this to save the additional lookup on line 78. That ended up making it two lines.

File: source/common/router/config_impl.cc (hunk 17 above)
Comment (tschroed) [inline source/common/router/config_impl.cc]: if (uniq) { return uniq.get() } else { return nullptr; } is the same as return uniq.get().

Comment (mattklein123) [inline source/common/router/config_impl.cc]: Ah sorry, my bad. I read incorrectly. Looks good.

File: source/common/common/trie.h [hunk 18]
```diff
@@ -0,0 +1,118 @@
+#pragma once
+
+namespace {
+// Because vector is efficient when removing elements from the back but paths
+// generally have the most significant components at the front, it's useful to
+// reverse the components vector before processing
+template <class T> inline void reverseComponentVector(std::vector<T>& vec) {
```
Comment (htuch) [inline source/common/common/trie.h]: Please split into trie.h and trie.cc. We shouldn't be dropping implementation detail helpers into unrelated code.

File: source/common/common/trie.h [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+  std::reverse(vec.begin(), vec.end());
```
Comment (htuch) [inline source/common/common/trie.h]: Plus, is this really worth wrapping? :)

File: source/common/common/trie.h [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+ * This is a general purpose trie. Keys and Values can be arbitrary classes
+ * provided they can be hashed. It supports both partial matches and exact
+ * matches.
+ *
+ * Note that Value must have a conversion constructor that will take nullptr.
+ */
+// TODO(tschroed): It's possible for Key to be a different type from the
+// internal path component. E.g. a uint32_t IPv4 prefix can be tokenized to
+// uint8_t components. Likewise a string representation of same.
+template <class Key, class Value, class Tokenizer> class TrieNode {
```
Comment (htuch) [inline source/common/common/trie.h]: VauePtr? or optional..

File: source/common/common/trie.h [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+   */
+  void emplace(const Key& key, const Value& value);
+
+  /**
+   * Lookup a value by key.
+   * @param key the path name to look up
+   * @return a std::pair consisting of the value if found and a boolean
+   *         indicating if the match was exact.
+   */
+  std::pair<Value*, bool> find(const Key& key) const;
```
Comment (htuch) [inline source/common/common/trie.h]: I wonder if there is any value in going the iterator route in making this even more consistent with STL maps. Probably not much value.

File: source/common/common/trie.h [hunk 22]
```diff
@@ -0,0 +1,118 @@
...
+// reverse the components vector before processing
+template <class T> inline void reverseComponentVector(std::vector<T>& vec) {
+  std::reverse(vec.begin(), vec.end());
+}
+} // namespace
+
+/**
+ * This is a general purpose trie. Keys and Values can be arbitrary classes
+ * provided they can be hashed. It supports both partial matches and exact
+ * matches.
```
Comment (htuch) [inline source/common/common/trie.h]: Please add a comment explaining the Value must be tokenizable by Tokenizer and that the trie uses these tokens for its path structure.

File: source/common/common/trie.h [hunk 23]
```diff
@@ -0,0 +1,118 @@
...
+  std::pair<Value*, bool> find(std::vector<Key>& path_components) const;
+  void emplace(std::vector<Key>& path_components, const Value& value);
+  void set_value(const Value& value) { value_ = value; }
+
+  // By holding pointers to the TriedNodes we don't have to worry that they're
+  // incomplete types.
+  std::unordered_map<Key, std::unique_ptr<TrieNode<Key, Value, Tokenizer>>> children_;
+  // value_ is being held as an opaque value and mutating it doesn't mutate the
+  // container.
+  mutable Value value_;
```
Comment (htuch) [inline source/common/common/trie.h]: Do we really need/want this? I understand the structure of the trie doesnt' change, but the trie as an overall value representation does. I'm thinking this would allow dangerous compiler optimizations.

File: source/common/common/trie.h [hunk 24]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+// TODO(tschroed): Make this iterative rather than recursive.
+template <class Key, class Value, class Tokenizer>
+std::pair<Value*, bool>
+TrieNode<Key, Value, Tokenizer>::find(std::vector<Key>& path_components) const {
+  if (path_components.empty()) {
+    // An exact match requires that we have a non-false value at the very end of
+    // our path traversal.
+    return std::make_pair(&value_, value_ ? true : false);
```
Comment (htuch) [inline source/common/common/trie.h]: We're returning a pointer to a pointer here? Why not just make Value the return type, since it basically need to be a pointer in any case (so it can have nullptr as a valid conversion). Also consider using optional to indicate match/no-match.

File: source/common/common/trie.h [hunk 25]
```diff
@@ -0,0 +1,118 @@
...
+  }
+  bool exact_match = false;
+  Value* value = &value_;
+  Key component(path_components.back());
+  path_components.pop_back();
+  auto it = children_.find(component);
+  if (it != children_.end()) {
+    const TrieNode<Key, Value, Tokenizer>* node = it->second.get();
+    std::pair<Value*, bool> retval = node->find(path_components);
```
Comment (htuch) [inline source/common/common/trie.h]: Did std::tie not help here?

File: source/common/common/utility.cc [hunk 26]
```diff
@@ -102,6 +102,30 @@ std::string StringUtil::subspan(const std::string& source, size_t start, size_t
...
+                                              const std::string& splitters) {
+  bool charmap[256] = {false};
```
Comment (htuch) [inline source/common/common/utility.cc]: Can we use sizeof(char) here instead of 256?

File: source/common/common/utility.cc [hunk 27]
```diff
@@ -102,6 +102,30 @@ std::string StringUtil::subspan(const std::string& source, size_t start, size_t
...
+  for (size_t i = 0; i < splitters.size(); i++) {
+    charmap[static_cast<uint8_t>(splitters[i])] = true;
```
Comment (htuch) [inline source/common/common/utility.cc]: Please cast to char, since std::string is defined as std::basic_string<char>.

File: source/common/common/utility.cc [hunk 28]
```diff
@@ -102,6 +102,30 @@ std::string StringUtil::subspan(const std::string& source, size_t start, size_t
...
+  }
+  size_t strlen = source.size();
+  std::vector<std::string> tokens;
+  size_t pos, i;
+  for (pos = 0, i = 0; i < strlen; i++) {
+    if (charmap[static_cast<uint8_t>(source[i])]) {
```
Comment (htuch) [inline source/common/common/utility.cc]: Ditto.

File: source/common/common/utility.cc [hunk 29]
```diff
@@ -102,6 +102,30 @@ std::string StringUtil::subspan(const std::string& source, size_t start, size_t
   return source.substr(start, end - start);
 }
 
+std::vector<std::string> StringUtil::tokenize(const std::string& source,
+                                              const std::string& splitters) {
+  bool charmap[256] = {false};
+  for (size_t i = 0; i < splitters.size(); i++) {
+    charmap[static_cast<uint8_t>(splitters[i])] = true;
+  }
+  size_t strlen = source.size();
```
Comment (htuch) [inline source/common/common/utility.cc]: It might be less scary not to use the same name as the C library function whose name shall never be spoken in polite company.

File: source/common/router/config_impl.h [hunk 30]
```diff
@@ -7,9 +7,21 @@
...
 #include "common/router/config_utility.h"
 
 namespace Router {
+// This is used to split a domain suffix into a series of path components (tokens) for placing in a
+// trie. E.g. ".foo.com" becomes {"com", ".", "foo", "."}. This is necessary because domain name
+// components proceed from most to least significant as right to left.
+struct ReverseDomainTokenizer {
+  std::vector<std::string> tokenize(const std::string& str) const {
+    std::vector<std::string> tokens(StringUtil::tokenize(str, "-."));
+    std::reverse(tokens.begin(), tokens.end());
```
Comment (htuch) [inline source/common/router/config_impl.h]: Is my mind playing tricks on me, or are we reversing both in the tokenizer and then in find/emplace again, thereby not reversing?

File: test/common/common/trie_test.cc [hunk 31]
```diff
@@ -0,0 +1,161 @@
...
+  std::vector<std::string> tokenize(const std::string& str) const {
+    std::vector<std::string> tokens(StringUtil::tokenize(str, "-."));
+    std::reverse(tokens.begin(), tokens.end());
+    return tokens;
+  }
+};
+
+TEST(TrieTest, DomainName) {
+  TrieNode<std::string, std::string*, ReverseDomainTokenizer> root;
+
```
Comment (htuch) [inline test/common/common/trie_test.cc]: Maybe add some tests for when the trie is completely empty.

File: test/common/common/trie_test.cc [hunk 32]
```diff
@@ -0,0 +1,161 @@
...
+  EXPECT_FALSE(vp.second);
+  vp = root.find("");
+  EXPECT_EQ(nullptr, *vp.first);
+  EXPECT_FALSE(vp.second);
+}
+
+class Uint64OctetTokenizer {
+public:
+  std::vector<uint64_t> tokenize(uint64_t num) const {
+    std::vector<uint64_t> octets{num >> 56 & 0xFF, num >> 48 & 0xFF, num >> 40 & 0xFF,
```
Comment (htuch) [inline test/common/common/trie_test.cc]: Nice!

File: source/common/common/trie.h (hunk 19 above)
Comment (mattklein123) [inline source/common/common/trie.h]: This was based on a previous comment from me of generally adding a comment and I think I mentioned a helper if that would help. I agree the wrapper is a bit excessive, I just care about comment clarity.

File: source/common/common/trie.h (hunk 23 above)
Comment (mattklein123) [inline source/common/common/trie.h]: I didn't comment on this before but I would agree with this. I would rather have it fully const for now.

File: source/common/common/trie.h (hunk 24 above)
Comment (mattklein123) [inline source/common/common/trie.h]: FYI this is from another comment of mine. There is no real requirement that Value be a pointer, in fact, it would be better if it didn't have to be. The only real requirement is that it is constructable from nullptr. I'm kind of fried right now, but optimally would do away with the nullptr requirement and just use Optional<> or a sentinel some type, then I think this make sense. Or have the template actually require shared_ptr<Value> if want it to only ever be a pointer. Would potentially like to consider some other solutions here but don't have brain power to offer anything right now.

Comment (htuch) [inline source/common/common/trie.h]: In addition to being nullptr constructable, Value needs to be testable (e.g. implement explicit operator bool() const) and have the properties that !Value(nullptr) always holds true.

Comment (mattklein123) [inline source/common/common/trie.h]: Right. I'm really not crazy about this nullptr business. What if we store an ```Optional<Value>```? Then the only requirement is value is default constructable. We could also then return an ```std::reference_wrapper<const Optional<Value>>```?

I think you might have suggested this above. The other option is to just make the template require shared_ptr as I mentioned elsewhere for now.

**Comment by mattklein123** [comment]
@tschroed could we finish this up and get this merged next week?

**Comment by tschroed** [comment]
Yes, this is on my list of things to do. Apologies for letting it languish.

**Comment by mattklein123** [comment]
Going to close this for now. Please reopen when we are ready to get this going again.
